lib/tree.rb

```diff
@@ -16,47 +16,154 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: log(n)
+  # Space Complexity: log(n)
+
+  def add_helper(key, value, current)
+
+    if current.key == key
+      current.value = value
+    elsif current.key > key
+      if current.left == nil
+        current.left = TreeNode.new(key,value)
+      else
+        add_helper(key, value, current.left)
+      end
+    else
+      if current.right == nil
+        current.right = TreeNode.new(key, value)
+      else
+        add_helper(key, value, current.right)
+      end
+    end
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add(key, value = nil)
+    if @root.nil?
+      @root = TreeNode.new(key,value) 
+      return @root
+    else
+      add_helper(key, value, @root)
+    end
+
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def find(key)
```

Comment:
👍 The time complexity is right if the tree is balanced and O(n) if not.  The space complexity (due to the call stack) is the same as the time complexity.